File: vane-admin/src/main/java/org/oddlama/vane/admin/ChatMessageFormatter.java
package org.oddlama.vane.admin;

import io.papermc.paper.chat.ChatRenderer;
import io.papermc.paper.event.player.AsyncChatEvent;
import net.kyori.adventure.audience.Audience;
import net.kyori.adventure.text.Component;
import net.kyori.adventure.text.format.NamedTextColor;
import org.bukkit.entity.Player;
import org.bukkit.event.EventHandler;
import org.bukkit.event.EventPriority;
import org.bukkit.event.player.PlayerJoinEvent;
import org.bukkit.event.player.PlayerKickEvent;
import org.bukkit.event.player.PlayerQuitEvent;
import org.oddlama.vane.annotation.lang.LangMessage;
import org.oddlama.vane.core.Listener;
import org.oddlama.vane.core.lang.TranslatedMessage;
import org.oddlama.vane.core.module.Context;

public class ChatMessageFormatter extends Listener<Admin> {

    @LangMessage
    private TranslatedMessage langPlayerChatFormat;

    @LangMessage
    private TranslatedMessage langPlayerJoin;

    @LangMessage
    private TranslatedMessage langPlayerKick;

    @LangMessage
    private TranslatedMessage langPlayerQuit;

    final ChatRenderer chatRenderer;

    public ChatMessageFormatter(Context<Admin> context) {
        super(
            context.group(
                "ChatMessageFormatter",
                "Enables custom formatting of chat messages like player chats and join / quit messages."
            )
        );
        // Create custom chat renderer
        chatRenderer = new ChatRenderer() {
            public Component render(
                final Player source,
                final Component sourceDisplayName,
                final Component message,
                final Audience viewer
            ) {
                // TODO more sophisticated formatting?
                final var who = sourceDisplayName.color(NamedTextColor.AQUA);
                return langPlayerChatFormat.strComponent(who, message);
            }
        };
    }

    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
    public void onPlayerChat(AsyncChatEvent event) {
        event.renderer(chatRenderer);
    }

    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
    public void onPlayerJoin(final PlayerJoinEvent event) {
        event.joinMessage(null);
        langPlayerJoin.broadcastServer(event.getPlayer().playerListName().color(NamedTextColor.GOLD));
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onPlayerKick(final PlayerKickEvent event) {
        // Bug in Spigot, doesn't do anything. But fixed in Paper since 1.17.
        // https://hub.spigotmc.org/jira/browse/SPIGOT-3034
        event.leaveMessage(Component.text(""));
        // message is handled in quit event
    }

    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    public void onPlayerQuit(final PlayerQuitEvent event) {
        event.quitMessage(null);
        if (event.getReason() == PlayerQuitEvent.QuitReason.KICKED) {
            langPlayerKick.broadcastServer(event.getPlayer().playerListName().color(NamedTextColor.GOLD));
        } else {
            langPlayerQuit.broadcastServer(event.getPlayer().playerListName().color(NamedTextColor.GOLD));
        }
    }
}

File: vane-admin/src/main/kotlin/org/oddlama/vane/admin/ChatMessageFormatter.kt
package org.oddlama.vane.admin

import io.papermc.paper.chat.ChatRenderer
import io.papermc.paper.event.player.AsyncChatEvent
import net.kyori.adventure.text.Component
import net.kyori.adventure.text.format.NamedTextColor
import org.bukkit.event.EventHandler
import org.bukkit.event.EventPriority
import org.bukkit.event.player.PlayerJoinEvent
import org.bukkit.event.player.PlayerKickEvent
import org.bukkit.event.player.PlayerQuitEvent
import org.oddlama.vane.annotation.lang.LangMessage
import org.oddlama.vane.core.Listener
import org.oddlama.vane.core.lang.TranslatedMessage
import org.oddlama.vane.core.module.Context

/**
 * Formats chat, join, and quit related messages.
 */
class ChatMessageFormatter(context: Context<Admin?>) : Listener<Admin?>(
    context.group(
        "ChatMessageFormatter",
        "Enables custom formatting of chat messages like player chats and join / quit messages."
    )
) {
    @LangMessage
    private val langPlayerChatFormat: TranslatedMessage? = null

    @LangMessage
    private val langPlayerJoin: TranslatedMessage? = null

    @LangMessage
    private val langPlayerKick: TranslatedMessage? = null

    @LangMessage
    private val langPlayerQuit: TranslatedMessage? = null

    /** Renderer used for chat message formatting. */
    private val chatRenderer = ChatRenderer { _, sourceDisplayName, message, _ ->
        val who = sourceDisplayName.color(NamedTextColor.AQUA)
        langPlayerChatFormat!!.strComponent(who, message)
    }

    /** Applies the custom chat renderer to async chat events. */
    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
    fun onPlayerChat(event: AsyncChatEvent) {
        event.renderer(chatRenderer)
    }

    /** Suppresses default join message and broadcasts configured join text. */
    @EventHandler(priority = EventPriority.LOWEST, ignoreCancelled = true)
    fun onPlayerJoin(event: PlayerJoinEvent) {
        event.joinMessage(null)
        langPlayerJoin!!.broadcastServer(event.player.playerListName().color(NamedTextColor.GOLD))
    }

    /**
     * Clears kick leave-message because quit handling already emits custom messaging.
     *
     * See Spigot issue SPIGOT-3034 for historical context.
     */
    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onPlayerKick(event: PlayerKickEvent) {
        event.leaveMessage(Component.empty())
    }

    /** Suppresses default quit message and broadcasts configured quit or kick text. */
    @EventHandler(priority = EventPriority.HIGHEST, ignoreCancelled = true)
    fun onPlayerQuit(event: PlayerQuitEvent) {
        event.quitMessage(null)
        if (event.reason == PlayerQuitEvent.QuitReason.KICKED) {
            langPlayerKick!!.broadcastServer(event.player.playerListName().color(NamedTextColor.GOLD))
        } else {
            langPlayerQuit!!.broadcastServer(event.player.playerListName().color(NamedTextColor.GOLD))
        }
    }
}
